metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/storage/BlockStorage.scala

```diff
@@ -0,0 +1,190 @@
+package io.iohk.metronome.hotstuff.service.storage
+
+import cats.implicits._
+import io.iohk.metronome.storage.{KVStore, KVStoreRead, KVCollection}
+import io.iohk.metronome.hotstuff.consensus.basic.{Agreement, Block}
+import scala.collection.immutable.Queue
+
+/** Storage for blocks that maintains parent-child relationships as well,
+  * to facilitate tree traversal and pruning.
+  *
+  * It is assumed that the application maintains some pointers into the tree
+  * where it can start traversing from, e.g. the last Commit Quorum Certificate
+  * would point at a block hash which would serve as the entry point.
+  */
+class BlockStorage[N, A <: Agreement: Block](
+    blockColl: KVCollection[N, A#Hash, A#Block],
+    childToParentColl: KVCollection[N, A#Hash, A#Hash],
+    parentToChildrenColl: KVCollection[N, A#Hash, Set[A#Hash]]
+) {
+  private implicit val kvn  = KVStore.instance[N]
+  private implicit val kvrn = KVStoreRead.instance[N]
+
+  /** Insert a block into the store, and if the parent still exists,
+    * then add this block to its children.
+    */
+  def put(block: A#Block): KVStore[N, Unit] = {
+    val blockHash  = implicitly[Block[A]].blockHash(block)
+    val parentHash = implicitly[Block[A]].parentBlockHash(block)
+
+    blockColl.put(blockHash, block) >>
+      childToParentColl.put(blockHash, parentHash) >>
+      parentToChildrenColl.put(blockHash, Set.empty) >>
+      parentToChildrenColl.update(parentHash, _ + blockHash)
+  }
+
+  /** Retrieve a block by hash, if it exists. */
+  def get(blockHash: A#Hash): KVStoreRead[N, Option[A#Block]] =
+    blockColl.read(blockHash)
+
+  /** Check whether a block is present in the tree. */
+  def contains(blockHash: A#Hash): KVStoreRead[N, Boolean] =
+    childToParentColl.read(blockHash).map(_.isDefined)
+
+  /** Check how many children the block has in the tree. */
+  private def childCount(blockHash: A#Hash): KVStoreRead[N, Int] =
+    parentToChildrenColl.read(blockHash).map(_.fold(0)(_.size))
+
+  /** Check whether the parent of the block is present in the tree. */
+  private def hasParent(blockHash: A#Hash): KVStoreRead[N, Boolean] =
+    childToParentColl.read(blockHash).flatMap {
+      case None             => KVStoreRead[N].pure(false)
+      case Some(parentHash) => contains(parentHash)
+    }
+
+  /** Check whether it's safe to delete a block.
+    *
+    * A block is safe to delete if doing so doesn't break up the tree
+    * into a forest, in which case we may have blocks we cannot reach
+    * by traversal, leaking space.
+    *
+    * This is true if the block has no children,
+    * or it has no parent and at most one child.
+    */
+  private def canDelete(blockHash: A#Hash): KVStoreRead[N, Boolean] =
+    (hasParent(blockHash), childCount(blockHash)).mapN {
+      case (_, 0)     => true
+      case (false, 1) => true
+      case _          => false
+    }
```

Comment:
Should we also check if the block or its ancestors have not been committed? Or is that check performed somewhere else?

Reply:
This one is a pure storage component, it doesn't have a concept of committed blocks. However, even if an ancestor is committed, it can be deleted if we want to free up space and our domain supports it, as it does in the case of Checkpointing: if we have a Q.C. for a checkpoint, we don't need the checkpoints before that, so we don't need to download the block history, and if we have it we can get rid of it after a while to keep storage space in check. 

Other applications may not have this property, in which case they wouldn't call pruning or deletion.